e2e node: Update runtime class handler skip logic

bobbypage opened the pull request:
Signed-off-by: David Porter <david@porter.me>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind bug

#### What this PR does / why we need it:

There are two runtime class tests which required the container runtime
config to include explicit configuration for `test-handler`. The current
logic skips these tests in non GCE environments. This skip is too strict
since the test is skipped in node e2e environments and in other
environments such as kind, which support running the test and also
configure `test-handler`.

Instead of skipping based on provider, add a new function
`NodeSupportsPreconfiguredRuntimeClassHandler` which examines the
underlying container runtime config and checks if the config includes
`test-handler`. The check is a bit brittle since it assumes container
runtime config paths, but it is a net improvement over skipping the test
entirely on non GCE environments.

This results in the test working in the common test environments, namely
GCE kube-up, node e2e, and kind.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #108804

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

bobbypage committed:
e2e node: Update runtime class handler skip logic

There are two runtime class tests which required the container runtime
config to include explicit configuration for `test-handler`. The current
logic skips these tests in non GCE environments. This skip is too strict
since the test is skipped in node e2e environments and in other
environments such as kind, which support running the test and also
configure `test-handler`.

Instead of skipping based on provider, add a new function
`NodeSupportsPreconfiguredRuntimeClassHandler` which examines the
underlying container runtime config and checks if the config includes
`test-handler`. The check is a bit brittle since it assumes container
runtime config paths, but it is a net improvement over skipping the test
entirely on non GCE environments.

This results in the test working in the common test environments, namely
GCE kube-up, node e2e, and kind.

Signed-off-by: David Porter <david@porter.me>

bobbypage commented:
/test ?

bobbypage commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-features

aojea commented on test/e2e/framework/node/runtimeclass/runtimeclass.go:
```diff
@@ -49,3 +54,25 @@ func NewRuntimeClassPod(runtimeClassName string) *v1.Pod {
+
+// Check if test-handler is configured. e.g. configured in kube-up on GCE:
+// see https://github.com/kubernetes/kubernetes/blob/eb729620c522753bc7ae61fc2c7b7ea19d4aad2f/cluster/gce/gci/configure-helper.sh#L3069-L3076
+func NodeSupportsPreconfiguredRuntimeClassHandler(ctx context.Context, f *framework.Framework) error {
```
this is super brittle for a conformance test, that may be used by who knows which new runtime that is not crio or containerd and not spawned with gci scripts

aojea replied:
or only runs on this `It` tests that are not conformance?

bobbypage replied:
I agree it's a bit brittle, but the goal is to remove the skip based on GCE, because the test is currently being skipped in node e2e even though the runtime handler is configured. Overall, I think it's a net improvement over skipping based on GCE... if the container runtime has it configured it is signal we can run the test.  

Additionally this makes the test run fine in kind for example, while today it is being skipped.


Regarding conformance, these tests are not run against conformance already due to what you mentioned (we can't add the test to conformance since the runtime class in the container runtime might be configured). 

The two tests are:

(e2e node test; not conformance) 

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/node/runtimeclass.go#L130-L133


(e2e cluster test; not conformance, but rather node feature)

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/common/node/runtimeclass.go#L83-L94

bobbypage commented:
/assign @SergeyKanzhelev @aojea

bobbypage commented:
This is ready to review now, updated the PR description. The test is now being run correctly in cluster e2e, node e2e and kind (while before it was being skipped in node e2e and kind).

Running in `pull-kubernetes-e2e-gce`:

<img width="1657" alt="image" src="https://user-images.githubusercontent.com/486583/214439140-6d4fc06a-a81b-43c4-9ac6-c195cac494db.png">

kind:

<img width="1647" alt="image" src="https://user-images.githubusercontent.com/486583/214439216-7446813b-3767-4e5a-b15d-fe2495225d30.png">

bobbypage commented:
But it is being skipped in `pull-kubernetes-node-e2e-containerd`, because the config there does not have [runtime-handler](https://github.com/kubernetes/test-infra/blob/master/jobs/e2e_node/containerd/config.toml) included:

![image](https://user-images.githubusercontent.com/486583/214440188-6dd00389-6df2-46de-8a7f-801baf3ea448.png)

bobbypage commented:
/retest

bobbypage commented:
/pull pull-kubernetes-cos-cgroupv2-containerd-node-e2e-features

bobbypage commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-features

bobbypage commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-features

aojea commented:
https://github.com/kubernetes/kubernetes/pull/115219#discussion_r1085981782
and is reflected as a comment in the code

/lgtm

The failing test is because of the CI issues with the images

SergeyKanzhelev reviewed (approved):
/kind cleanup
/kind failing-test
/lgtm
/approve

aojea commented:
/approve

bobbypage commented:
/retest

bobbypage commented:
/hold

bobbypage commented:
Holding to take a look at the failure above

bobbypage committed:
test: Update Host Exec to support node e2e

Node E2E tests do not run a scheduler, so the host exec pod must have
the `spec.nodeName` set explicitly.

Signed-off-by: David Porter <david@porter.me>

bobbypage commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-features

bobbypage commented:
Pushed up another commit (6cb022ff4a3808b9517e965ed1fb7da063e164ae) to fix the issue with the node e2e test, now all the tests are green and this should be good to go.

/unhold

bobbypage commented:
/cc @SergeyKanzhelev @aojea 

can you please re-lgtm?

SergeyKanzhelev reviewed (approved):
/lgtm

bobbypage commented:
/retest

hakman commented:
@bobbypage After merging this PR, the RuntimeClass tests are failing for the kOps project. Could you take a look? Thanks!
https://testgrid.k8s.io/kops-network-plugins#kops-aws-cni-calico-deb11

bobbypage commented:
@hakman

hmm, I see in the logs:
```
  Jan 31 06:17:33.851: INFO: At 2023-01-31 06:12:32 +0000 UTC - event for test-runtimeclass-runtimeclass-9011-non-conflict-runtimeclwr5r7: {kubelet i-0bdd5528629f2fd5a} FailedCreatePodSandBox: Failed to create pod sandbox: rpc error: code = Unknown desc = failed to get sandbox runtime: no runtime for "test-handler" is configured
```

but the test is suppose to skip if `test-handler` is not configured... What is the containerd config.toml config used in this job?

hakman commented:
@bobbypage There is no `/etc/containerd/config.toml`. At the moment we use `/etc/containerd/config-kops.toml`.

bobbypage commented:
ah, that explains it, we should return an error if there is no `/etc/containerd/config.toml`. will open pr for it.

bobbypage commented:
I opened PR here - https://github.com/kubernetes/kubernetes/pull/115437
